pandas_gbq/gbq.py

```diff
@@ -956,61 +797,148 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
 
     """
 
-    _test_google_api_imports()
-
-    if not project_id:
-        raise TypeError("Missing required parameter: project_id")
-
     if dialect not in ('legacy', 'standard'):
         raise ValueError("'{0}' is not valid for dialect".format(dialect))
+    if configuration and any(key in configuration for key in
+                             ["query", "copy", "load", "extract"]):
+        raise ValueError("New API handles configuration settings differently")
+
+    def _wait_for_job(job):
+        while True:
+            job.reload()  # Refreshes the state via a GET request.
+            if job.state == 'DONE':
+                if job.error_result:
+                    raise RuntimeError(job.errors)
+                return
+            time.sleep(1)
 
-    connector = GbqConnector(
-        project_id, reauth=reauth, verbose=verbose, private_key=private_key,
-        dialect=dialect, auth_local_webserver=auth_local_webserver)
-    schema, pages = connector.run_query(query, **kwargs)
-    dataframe_list = []
-    while len(pages) > 0:
-        page = pages.pop()
-        dataframe_list.append(_parse_data(schema, page))
-
-    if len(dataframe_list) > 0:
-        final_df = concat(dataframe_list, ignore_index=True)
+    if credentials is None:
+        credentials = GbqConnector(project_id=project_id,
+                                   reauth=reauth,
+                                   auth_local_webserver=auth_local_webserver,
+                                   private_key=private_key).credentials
+    client = bigquery.Client(project=project_id, credentials=credentials)
+
+    def _set_common_query_settings(query_job):
+        if dialect == 'legacy':
+            query_job.use_legacy_sql = True
+        elif dialect == 'standard':
+            query_job.use_legacy_sql = False
+
+        if configuration:
+            for setting, value in configuration.items():
+                setattr(query_job, setting, value)
+        return query_job
+
+    def sync_query():
```

Comment:
What are these `sync_query` and `async_query` functions doing here?

Only the `async_query` method should be used. `sync_query` was misnamed as it does still start a query job behind the scenes. If the call times out, you still need to poll for the job to be complete, just like with the "async" method, so you might as well remove it because it doesn't gain you anything.

Reply:
Initially didn't see in the docs a way to set timeout in async query, plus the docs (on the main page no less http://google-cloud-python.readthedocs.io/en/latest/bigquery/usage.html#querying-data-synchronous :) had sample for how to do so with sync query. I see in code that I can set it by passing `timeout` to `result()` in an async_query, but when I test it doesn't work. Can you confirm that this is the right way to specify a timeout?

Reply:
Yes, timeout on `result()` should work. If it does not, please file a bug with the reproducible error at https://github.com/GoogleCloudPlatform/google-cloud-python/issues

Reply:
Issue raised here: https://github.com/GoogleCloudPlatform/google-cloud-python/issues/4135. Until I either get clarification or workaround, I'll leave sync query in for now in order to allow for timing out queries. If/when timeout works for async queries, I can easily strip out all the sync query code and implement timeout for async, hope that makes sense.

Reply:
You're right. There's a bug in how the timeout is handled. I'll work to fix it in a subsequent version of `google-cloud-bigquery`, but you're right that this workaround is required.

Reply:
Fixed in [0.28.0](https://github.com/GoogleCloudPlatform/google-cloud-python/releases/tag/bigquery-0.28.0) (Though that includes a bunch of other breaking changes, so I understand if you don't want to update this PR. Sorry about that!)

Reply:
Yeah, going to try and get this working version merged before having to deal with new breaking changes, thanks for the update! :)